src/targeting/build-page-targeting.ts

```diff
@@ -0,0 +1,151 @@
+import type { Participations } from '@guardian/ab-core';
+import { cmp } from '@guardian/consent-management-platform';
+import type { ConsentState } from '@guardian/consent-management-platform/dist/types';
+import type { CountryCode } from '@guardian/libs';
+import { getCookie, isString } from '@guardian/libs';
+import type { False, True } from '../types';
+import type { ContentTargeting } from './content';
+import { getContentTargeting } from './content';
+import type { AdManagerGroup, Frequency } from './personalised';
+import { getPersonalisedTargeting } from './personalised';
+import type { SessionTargeting } from './session';
+import { getSessionTargeting } from './session';
+import type { SharedTargeting } from './shared';
+import { getSharedTargeting } from './shared';
+import type { ViewportTargeting } from './viewport';
+import { getViewportTargeting } from './viewport';
+
+type PartialWithNulls<T> = { [P in keyof T]?: T[P] | null };
+
+type TrueOrFalse = True | False;
+
+type PageTargeting = PartialWithNulls<
+	{
+		ab: string[];
+		af: 't'; // Ad Free
+		amtgrp: AdManagerGroup;
+		at: string; // Ad Test
+		bp: 'mobile' | 'tablet' | 'desktop'; // BreakPoint
+		cc: CountryCode; // Country Code
+		cmp_interaction: string;
+		consent_tcfv2: string;
+		dcre: TrueOrFalse; // DotCom-Rendering Eligible
+		fr: Frequency; // FRequency
+		inskin: TrueOrFalse; // InSkin
+		pa: TrueOrFalse; // Personalised Ads consent
+		permutive: string[]; // predefined segment values
+		pv: string; // ophan Page View id
+		rdp: string;
+		ref: string; // REFerrer
+		rp: 'dotcom-rendering' | 'dotcom-platform'; // Rendering Platform
+		s: string; // site Section
+		sens: TrueOrFalse; // SenSitive
+		si: TrueOrFalse; // Signed In
+		skinsize: 'l' | 's';
+		urlkw: string[]; // URL KeyWords
+		vl: string; // Video Length
+
+		// And more
+		[_: string]: string | string[];
+	} & SharedTargeting
+>;
+
+const filterEmptyValues = (pageTargets: Record<string, unknown>) => {
+	const filtered: Record<string, string | string[]> = {};
+	for (const key in pageTargets) {
+		const value = pageTargets[key];
+		if (isString(value)) {
+			filtered[key] = value;
+		} else if (
+			Array.isArray(value) &&
+			value.length > 0 &&
+			value.every(isString)
+		) {
+			filtered[key] = value;
+		}
+	}
+	return filtered;
+};
+
+const buildPageTargeting = (
+	consentState: ConsentState,
+	adFree: boolean,
+	countryCode: CountryCode,
+	clientSideParticipations: Participations,
+): PageTargeting => {
+	if (
+		!window.guardian.config?.page ||
+		!window.guardian.config.ophan ||
+		window.guardian.config.isDotcomRendering === undefined
+	) {
+		return {};
+	}
```

Comment:
This doesn't feel great, is there a better way?

Reply:
Is this new functionality?

I'm not sure how 'fatal' these conditions are. Maybe there's an argument to say we should try to send as much targeting as we can even if some of the properties are empty? Might be worth asking Alex.

Reply:
There's some mismatches between what [Frontend considers present](https://github.com/guardian/frontend/blob/main/static/src/javascripts/types/global.d.ts#L376) on the window and what [commercial-core does](https://github.com/guardian/commercial-core/blob/main/src/global.ts#L23) - could we fix our types up to match what is present in Frontend?

Reply:
I would have thought its better in general to try and send as much targeting as possible even if some properties are missing in order to maximise fulfilment.

But putting that aside:

[Dotcom](https://github.com/guardian/dotcom-rendering/blob/89706d26c0298b19449c49a9a79681b8cbe3af98/dotcom-rendering/src/model/window-guardian.ts#L9) and [Frontend](https://github.com/guardian/frontend/blob/1aaed4efc97a181f06b7a4a29d7ac528d01db7aa/static/src/javascripts/types/global.d.ts#L79) consider `guardian.config.page` (and `guardian.config.isDotcomRendering`) to be present.

[Dotcom](https://github.com/guardian/dotcom-rendering/blob/89706d26c0298b19449c49a9a79681b8cbe3af98/dotcom-rendering/src/model/window-guardian.ts#L33) and [Frontend](https://github.com/guardian/frontend/blob/1aaed4efc97a181f06b7a4a29d7ac528d01db7aa/static/src/javascripts/types/global.d.ts#L74) consider `guardian.config.ophan` to be present.

[Dotcom](https://github.com/guardian/dotcom-rendering/blob/89706d26c0298b19449c49a9a79681b8cbe3af98/dotcom-rendering/window.guardian.d.ts#L24) considers `guardian.ophan` to be optional whereas [Frontend](https://github.com/guardian/frontend/blob/1aaed4efc97a181f06b7a4a29d7ac528d01db7aa/static/src/javascripts/types/global.d.ts#L375) does not.

So as long as we use `guardian.config.ophan.pageViewId` I think we can rely on Dotcom and Frontend getting the types right and remove this if block?

Reply:
Thanks both, that is really helpful. I think fixing the inconsistencies Chris mentioned should just fix this and remove the need for the guard statement

Reply:
I've resolved this [here](https://github.com/guardian/commercial-core/pull/686/commits/ba47f077093492ec8a4897fa10a142d6452e6073), should be good now!